escrow/free_call.go

```diff
@@ -0,0 +1,151 @@
+	log "github.com/sirupsen/logrus"
+)
+
+type lockingFreeCallUserService struct {
+	storage        *FreeCallUserStorage
+	locker         Locker
+	replicaGroupID func() ([32]byte, error)
+}
+
+func NewFreeCallUserService(
+	storage *FreeCallUserStorage,
+
+	locker Locker,
+	groupIdReader func() ([32]byte, error)) FreeCallUserService {
+
+	return &lockingFreeCallUserService{
+		storage:        storage,
+		locker:         locker,
+		replicaGroupID: groupIdReader,
+	}
+}
+
+func (h *lockingFreeCallUserService) FreeCallUserUsage(key *FreeCallUserKey) (freeCallUser *FreeCallUserData, ok bool, err error) {
+	freeCallUser, ok, err = h.storage.Get(key)
+	if err != nil {
+		return
+	}
+	if !ok {
+		groupId, err := h.replicaGroupID()
+		if err != nil {
+			return nil, false, err
+		}
+		return &FreeCallUserData{UserId: key.UserId, OrgId: key.OrganizationId, ServiceId: key.ServiceId, GroupID: groupId, FreeCallsMade: 0}, true, nil
+	}
+	return
+}
+
+func (h *lockingFreeCallUserService) ListFreeCallUsers() (users []*FreeCallUserData, err error) {
+	return h.storage.GetAll()
+}
+
+type freeCallTransaction struct {
+	payment      FreeCallPayment
+	freeCallUser *FreeCallUserData
+	service      *lockingFreeCallUserService
+	lock         Lock
+}
+
+func (transaction *freeCallTransaction) String() string {
+	return fmt.Sprintf("{FreeCallPayment: %v, FreeCallUser: %v}", transaction.payment, transaction.freeCallUser)
+}
+
+func (transaction *freeCallTransaction) FreeCallUser() *FreeCallUserData {
+	return transaction.freeCallUser
+}
+
+func (h *lockingFreeCallUserService) StartFreeCallUserTransaction(payment *FreeCallPayment) (transaction FreeCallTransaction, err error) {
+	groupId, err := h.replicaGroupID()
+	userKey := &FreeCallUserKey{UserId: payment.UserId, OrganizationId: payment.OrganizationId,
+		ServiceId: payment.ServiceId, GroupID: blockchain.BytesToBase64(groupId[:])}
+	freeCallUserData, ok, err := h.FreeCallUserUsage(userKey)
+	if err != nil {
+		return nil, NewPaymentError(Internal, "payment freeCallUserData error:"+err.Error())
+	}
+	if !ok {
+		log.Warn("Payment freeCallUserData not found")
+		return nil, NewPaymentError(Unauthenticated, "payment freeCallUserData \"%v\" not found", userKey)
+	}
+	if err != nil {
+		return nil, NewPaymentError(Internal, "cannot get mutex for user: %v", payment.UserId)
+	}
+
+	lock, ok, err := h.locker.Lock(userKey.String())
+	if err != nil {
+		return nil, NewPaymentError(Internal, "cannot get mutex for user: %v", userKey)
+	}
+	if !ok {
+		return nil, NewPaymentError(FailedPrecondition, "another transaction on this user: %v is in progress", userKey)
+	}
+	defer func(lock Lock) {
+		if err != nil {
+			e := lock.Unlock()
+			if e != nil {
+				log.WithError(e).WithField("userKey", userKey).WithField("err", err).Error("Transaction is cancelled because of err, but freeCallUserData cannot be unlocked. All other transactions on this freeCallUserData will be blocked until unlock. Please unlock freeCallUserData manually.")
```

Comment:
Should we send a notification in this case? if this code executes the user is essentially completely blocked so we should make it easy for the service developer to remedy this

Reply:
we can do this for the channel locks too